File: 프로그래머스/LV2/코딩테스트 연습/위클리 챌린지/교점에 별 만들기/교점에 별 만들기.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <cmath>
using namespace std;

vector<string> solution(vector<vector<int>> line)
{
    vector<string> answer;
    vector<long long int> xs, ys;
    int line_len = line.size();
    int row, col;
    long long int min_x, max_x, min_y, max_y;
    double x, y, xn, yn, d;

    for (int i = 0; i < line_len; i++)
    {
        for (int j = i+1; j < line_len; j++)
        {
            d = (double)((long long int)line[i][0] * (long long int)line[j][1]) - (double)((long long int)line[i][1] * (long long int)line[j][0]);
            if (!d)continue;
            xn = (double)((long long int)line[i][1] * (long long int)line[j][2]) - (double)((long long int)line[i][2] * (long long int)line[j][1]);
            yn = (double)((long long int)line[i][2] * (long long int)line[j][0]) - (double)((long long int)line[i][0] * (long long int)line[j][2]);
            x = xn / d;
            y = yn / d;
            if ((x - floor(x) == 0) && (y - floor(y) == 0))
            {
                xs.push_back(floor(x));
                ys.push_back(floor(y));
            }
        }
    }

    min_x = *min_element(xs.begin(), xs.end());
    max_x = *max_element(xs.begin(), xs.end());
    min_y = *min_element(ys.begin(), ys.end());
    max_y = *max_element(ys.begin(), ys.end());
    for (int i = 0; i < xs.size(); i++)xs[i] -= min_x;
    for (int i = 0; i < ys.size(); i++)ys[i] = max_y - ys[i];
    row = max_y - (min_y - 1);
    col = max_x - (min_x - 1);

    for (int r = 0; r < row; r++)
    {
        answer.push_back("");
        for (int c = 0; c < col; c++)answer[r] += '.';
    }
    for (int i = 0; i < xs.size(); i++)answer[ys[i]][xs[i]] = '*';
    return answer;
}

File: 프로그래머스/LV2/코딩테스트 연습/위클리 챌린지/교점에 별 만들기/교점에 별 만들기.py
def solution(line):
    answer = []
    xs = []
    ys = []
    line_len = len(line)
    for i in range(line_len):
        for j in range(i+1, line_len):
            d = line[i][0]*line[j][1] - line[i][1]*line[j][0]
            if d == 0:
                continue
            xn = line[i][1]*line[j][2] - line[i][2]*line[j][1]
            yn = line[i][2]*line[j][0] - line[i][0]*line[j][2]
            x = xn / d
            y = yn / d
            if x.is_integer() and y.is_integer():
                xs.append(int(x))
                ys.append(int(y))

    min_x,max_x,min_y,max_y = min(xs), max(xs), min(ys), max(ys)
    xs = [x - min_x for x in xs]
    ys = [max_y - y for y in ys]

    row = max_y - (min_y-1)
    col = max_x - (min_x-1)

    for _ in range(row):
        answer.append(['.'] * col)
    for y, x in zip(ys,xs):
        answer[y][x] = '*'
    for r in range(row):
        answer[r] = "".join(answer[r])
    return answer